ambry-clustermap/src/main/java/com/github/ambry/clustermap/InstanceConfigToDataNodeConfigAdapter.java

```diff
@@ -33,82 +38,146 @@
+     */
+    DataNodeConfig convert(InstanceConfig instanceConfig) {
+      int schemaVersion = getSchemaVersion(instanceConfig);
+      if (schemaVersion != 0) {
+        LOGGER.warn("Unknown InstanceConfig schema version {} in {}. Ignoring.", schemaVersion, instanceConfig);
+        return null;
+      }
+      DataNodeConfig dataNodeConfig = new DataNodeConfig(instanceConfig.getInstanceName(), instanceConfig.getHostName(),
+          Integer.parseInt(instanceConfig.getPort()), getDcName(instanceConfig), getSslPortStr(instanceConfig),
+          getHttp2PortStr(instanceConfig), getRackId(instanceConfig), getXid(instanceConfig));
+      dataNodeConfig.getSealedReplicas().addAll(getSealedReplicas(instanceConfig));
+      dataNodeConfig.getStoppedReplicas().addAll(getStoppedReplicas(instanceConfig));
+      dataNodeConfig.getDisabledReplicas().addAll(getDisabledReplicas(instanceConfig));
+      instanceConfig.getRecord().getMapFields().forEach((mountPath, diskProps) -> {
+        if (diskProps.get(DISK_STATE) == null) {
+          // Check if this map field actually holds disk properties, since we can't tell from just the field key (the
+          // mount path with no special prefix). There may be extra fields when Helix controller adds partitions in ERROR
+          // state to InstanceConfig.
+          LOGGER.warn("{} field does not contain disk info on {}. Skip it and continue on next one.", mountPath,
+              instanceConfig.getInstanceName());
+        } else {
+          DataNodeConfig.DiskConfig disk = new DataNodeConfig.DiskConfig(
+              diskProps.get(DISK_STATE).equals(AVAILABLE_STR) ? HardwareState.AVAILABLE : HardwareState.UNAVAILABLE,
+              Long.parseLong(diskProps.get(DISK_CAPACITY_STR)));
+          String replicasStr = diskProps.get(REPLICAS_STR);
+          if (!Utils.isNullOrEmpty(replicasStr)) {
+            for (String replicaStr : replicasStr.split(REPLICAS_DELIM_STR)) {
+              String[] replicaStrParts = replicaStr.split(REPLICAS_STR_SEPARATOR);
+              // partition name and replica name are the same.
+              String partitionName = replicaStrParts[0];
+              long replicaCapacity = Long.parseLong(replicaStrParts[1]);
+              String partitionClass =
+                  replicaStrParts.length > 2 ? replicaStrParts[2] : clusterMapConfig.clusterMapDefaultPartitionClass;
+              disk.getReplicaConfigs()
+                  .put(partitionName, new DataNodeConfig.ReplicaConfig(replicaCapacity, partitionClass));
+            }
           }
+          dataNodeConfig.getDiskConfigs().put(mountPath, disk);
         }
-        dataNodeConfig.getDiskConfigs().put(mountPath, disk);
+      });
+      return dataNodeConfig;
+    }
+
+    /**
+     * @param dataNodeConfig the {@link DataNodeConfig} to convert into an {@link InstanceConfig}.
+     * @return the {@link InstanceConfig}.
+     */
+    InstanceConfig convert(DataNodeConfig dataNodeConfig) {
```

Comment:
A quick check, do we have any test cases cover this method?

Reply:
If you would like, I can make unit tests specifically for the converter. Currently they are tested using the DataNodeConfigSource.set and get methods, which call these functions.

Reply:
As long as we have test call these functions, I'm fine with it. Thanks for clarify.